File: ga/4omini_result_2024_7/result/1968F_cpp_mini/13_1968F.cpp
use std::io::{self, BufRead};
use std::cmp::Ordering;

#[derive(Debug, Copy, Clone)]
struct Pair {
    x: i64, // first element of the pair (XOR)
    index: usize, // index in the original array
}

// Implementing a function to compare pairs by their XOR value
impl PartialOrd for Pair {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        match self.x.cmp(&other.x) {
            Ordering::Equal => self.index.cmp(&other.index).then(Some(Ordering::Equal)),
            ord => Some(ord),
        }
    }
}

impl PartialEq for Pair {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.index == other.index
    }
}

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();

    let t: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    
    for _ in 0..t {
        let line = lines.next().unwrap().unwrap();
        let mut inputs = line.trim().split_whitespace().map(|s| s.parse::<usize>().unwrap());
        let n = inputs.next().unwrap();
        let q = inputs.next().unwrap();

        let mut arr = Vec::new();
        let line = lines.next().unwrap().unwrap();
        for number in line.trim().split_whitespace() {
            arr.push(number.parse::<i64>().unwrap());
        }

        let mut x = vec![Pair { x: 0, index: 0 }; n + 1];
        let mut x2 = vec![0; n + 1];

        for i in 1..=n {
            x[i].x = x[i - 1].x ^ arr[i - 1];
            x[i].index = i;
        }

        for i in 0..=n {
            x2[i] = x[i].x;
        }

        x.sort_by(|a, b| a.partial_cmp(b).unwrap());

        for _ in 0..q {
            let line = lines.next().unwrap().unwrap();
            let mut lr = line.trim().split_whitespace().map(|s| s.parse::<usize>().unwrap());
            let l = lr.next().unwrap();
            let r = lr.next().unwrap();

            let c = x2[r] ^ x2[l - 1];
            if c == 0 {
                println!("YES");
                continue;
            }

            let target = c ^ x2[l - 1];
            let i1 = x.binary_search(&Pair { x: target, index: l }).unwrap_err();
            
            if i1 == n + 1 || x[i1].index > r || x[i1].x != target {
                println!("NO");
                continue;
            }

            let i2 = x.binary_search(&Pair { x: x2[l - 1], index: x[i1].index + 1 }).unwrap_err();
            
            if i2 == n + 1 || x[i2].index > r || x[i2].x != x2[l - 1] {
                println!("NO");
            } else {
                println!("YES");
            }
        }
    }
}

File: ga/glm4_result_2024_6/result/1968F_cpp_glm/13_1968F.rs
use std::io::{self, BufRead};

fn solve() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    for _ in 0..t {
        let nq: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();
        let n = nq[0];
        let q = nq[1];
        let arr: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();
        let mut x = vec![(0, 0); n + 1];
        let mut x2 = vec![0; n + 1];
        x[0] = (0, 0);
        for i in 1..=n {
            x[i] = (x[i - 1].0 ^ arr[i - 1], i);
            x2[i] = x[i].0;
        }
        x.sort_by(|a, b| a.0.cmp(&b.0));

        for _ in 0..q {
            let lr: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();
            let l = lr[0];
            let r = lr[1];
            let c = x2[r] ^ x2[l - 1];
            if c == 0 {
                println!("YES");
                continue;
            }
            let i1 = x.binary_search_by(|&(ref a, ref b)| {
                if a == &c ^ x2[l - 1] {
                    b.cmp(&l)
                } else {
                    a.cmp(&c ^ x2[l - 1])
                }
            }).unwrap_or_else(|i| i);

            if i1 == n + 1 || x[i1].1 > r || x[i1].0 != (c ^ x2[l - 1]) {
                println!("NO");
                continue;
            }
            let i2 = x.binary_search_by(|&(ref a, ref b)| {
                if a == &x2[l - 1] {
                    b.cmp(&(x[i1].1 + 1))
                } else {
                    a.cmp(&x2[l - 1])
                }
            }).unwrap_or_else(|i| i);

            if i2 == n + 1 || x[i2].1 > r || x[i2].0 != x2[l - 1] {
                println!("NO");
            } else {
                println!("YES");
            }
        }
    }
}

fn main() {
    solve();
}